cmd/secrets/secret_helper.go

```diff
@@ -53,105 +83,114 @@ type secretsRequest struct {
+		return nil, errors.New("failed to unmarshal json input")
 	}
 
 	version := splitVersion(request.Version)
 	compatVersion := splitVersion(s.PayloadVersion)
 	if version[0] != compatVersion[0] {
-		return fmt.Errorf("incompatible protocol version %q", request.Version)
+		return nil, fmt.Errorf("incompatible protocol version %q", request.Version)
 	}
 
 	if len(request.Secrets) == 0 {
-		return errors.New("no secrets listed in input")
+		return nil, errors.New("no secrets listed in input")
 	}
 
-	response := map[string]s.Secret{}
-	for _, name := range request.Secrets {
-		response[name] = readSecret(filepath.Join(dir, name))
-	}
+	return request.Secrets, nil
+}
 
-	out, err := json.Marshal(response)
+func writeFetchedSecrets(w io.Writer, fetchedSecrets map[string]s.Secret) error {
+	out, err := json.Marshal(fetchedSecrets)
 	if err != nil {
 		return err
 	}
+
 	_, err = w.Write(out)
 	return err
 }
 
-func splitVersion(ver string) []string {
-	return strings.SplitN(ver, ".", 2)
-}
+func readSecretsFromFile(secrets []string, dir string) map[string]s.Secret {
+	res := make(map[string]s.Secret)
 
-func readSecret(path string) s.Secret {
-	value, err := readSecretFile(path)
-	var errMsg string
-	if err != nil {
-		errMsg = err.Error()
+	secretsProvider := providers.NewFileProvider(dir)
+	for _, secretID := range secrets {
+		res[secretID] = secretsProvider.Get(secretID)
 	}
-	return s.Secret{Value: value, ErrorMsg: errMsg}
+
+	return res
 }
 
-func readSecretFile(path string) (string, error) {
-	fi, err := os.Lstat(path)
-	if err != nil {
-		if os.IsNotExist(err) {
-			return "", errors.New("secret does not exist")
-		}
-		return "", err
-	}
+func readSecretsUsingPrefixes(secrets []string, selector *providers.Selector, rootPath string) map[string]s.Secret {
+	res := make(map[string]s.Secret)
 
-	if fi.Mode()&os.ModeSymlink != 0 {
-		// Ensure that the symlink is in the same dir
-		target, err := os.Readlink(path)
+	for _, secretID := range secrets {
+		prefix, id, err := parseSecretWithPrefix(secretID, rootPath)
 		if err != nil {
-			return "", fmt.Errorf("failed to read symlink target: %v", err)
-		}
-
-		dir := filepath.Dir(path)
-		if !filepath.IsAbs(target) {
-			target, err = filepath.Abs(filepath.Join(dir, target))
-			if err != nil {
-				return "", fmt.Errorf("failed to resolve symlink absolute path: %v", err)
-			}
+			res[secretID] = s.Secret{Value: "", ErrorMsg: err.Error()}
+			continue
 		}
 
-		dirAbs, err := filepath.Abs(dir)
+		secretsProvider, err := selector.GetProvider(prefix)
 		if err != nil {
-			return "", fmt.Errorf("failed to resolve absolute path of directory: %v", err)
+			res[secretID] = s.Secret{Value: "", ErrorMsg: err.Error()}
+			continue
 		}
 
-		if !filepath.HasPrefix(target, dirAbs) {
-			return "", fmt.Errorf("not following symlink %q outside of %q", target, dir)
-		}
-	}
-	fi, err = os.Stat(path)
-	if err != nil {
-		return "", err
+		res[secretID] = secretsProvider.Get(id)
```

Comment:
I still feel that all the `selector` logic, the struct wrapper around `readSecretFile` and `readKubernetesSecret` is an abstraction around a simple if/else here.

In my opinion, two simple function and a if-else/switch here would be simpler to test and maintain.